src/Functions/FunctionsJSON.h

```diff
@@ -286,27 +288,81 @@ class FunctionJSON : public IFunction, WithContext
     String getName() const override { return Name::name; }
     bool isVariadic() const override { return true; }
     size_t getNumberOfArguments() const override { return 0; }
+
+    // Both NULL and JSON NULL should generate NULL value.
+    bool useDefaultImplementationForNulls() const override { return false; }
+
     bool useDefaultImplementationForConstants() const override { return true; }
     bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
 
     DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
     {
+        NullPresence null_presence = getNullPresense(arguments);
+
+        if (null_presence.has_null_constant)
+        {
+            return makeNullable(std::make_shared<DataTypeNothing>());
+        }
+        if (null_presence.has_nullable)
+        {
+            auto nested_columns = createBlockWithNestedColumns(arguments);
+            auto return_type = Impl<DummyJSONParser>::getReturnType(Name::name, nested_columns);
+            return makeNullable(return_type);
+        }
         return Impl<DummyJSONParser>::getReturnType(Name::name, arguments);
     }
 
     ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const override
     {
+        NullPresence null_presence = getNullPresense(arguments);
+        if (null_presence.has_null_constant)
+        {
+            // Default implementation for nulls returns null result for null arguments,
+            // so the result type must be nullable.
+            if (!result_type->isNullable())
+                throw Exception(ErrorCodes::LOGICAL_ERROR,
+                                "Function {} with Null argument and default implementation for Nulls "
+                                "is expected to return Nullable result, got {}", result_type->getName());
```

Comment:
It looks like this error should be for the more general case `null_presence.has_nullable==true`, not only for `null_presence.has_null_constant==true`.

Reply:
The exception is useless. It's not meant to catch any bug... I'll rewrite.